docs/src/main/sphinx/develop/client-protocol.rst

```diff
@@ -0,0 +1,257 @@
+
+The ``status`` field of the JSON document is for human consumption only, and
+provides a hint about the query's state on the server.  It is not synchronized
+with the server's query state and should not be used to tell if the query
+is finished.
+
+Important ``QueryResults`` attributes
+=====================================
+
+The most important attributes of the ``QueryResults`` JSON document returned by the REST API
+endpoints are listed in this table.  Refer to the class ``QueryResults`` for more details.
+
+.. list-table:: ``QueryResults attributes``
+  :widths: 25, 55
+  :header-rows: 1
+
+  * - Attribute
+    - Description
+  * - ``id``
+    - The ID of the query.
+  * - ``nextUri``
+    - If present, the URL that should be used in subsequent ``GET`` or
+      ``DELETE`` requests.  If not present, the query is complete or
+      ended in error.
+  * - ``columns``
+    - A list of the names and types of the columns returned by the query.
+  * - ``data``
+    - The ``data`` attribute contains a list of the rows returned by the
+      query request.  Each row is itself a list that holds values of the
+      columns in the row, in the order specified by the ``columns``
+      attribute.
+  * - ``updateType``
+    - A human-readable string representing the operation.  For a
+      ``CREATE TABLE`` request, the ``updateType`` will be
+      "CREATE TABLE"; for ``SET SESSION`` it will be "SET SESSION"; etc.
+  * - ``error``
+    - If query failed, the ``error`` attribute will contain JSON for a
+      ``QueryError`` object.  That object contains a ``message``, an
+      ``errorCode`` and other information about the error.  See the
+      ``QueryError`` class for more details.
+
+
+``QueryResults`` diagnostic attributes
+======================================
+
+These ``QueryResults`` data members may be useful in tracking down problems:
+
+.. list-table:: ``QueryResults diagnostic attributes``
+  :widths: 20, 20, 40
+  :header-rows: 1
+
+  * - Attribute
+    - Type
+    - Description
+  * - ``queryError``
+    - ``QueryError``
+    - Non-null only if the query resulted in an error.
+  * - ``failureInfo``
+    - ``FailureInfo``
+    - ``failureInfo`` has detail on the reason for the failure, including
+      a stack trace, and ``FailureInfo.errorLocation``, providing the
+      query line number and column number where the failure was detected.
+  * - ``warnings``
+    - ``List<TrinoWarning>``
+    - A usually-empty list of warnings.
+  * - ``statementStats``
+    - ``StatementStats``
+    - A class containing statistics about the query execution.  Of
+      particular interest is ``StatementStats.rootStage``, of type
+      ``StageStats``, providing statistics on the execution of each of
+      the stages of query processing.
+
+Client request headers
+======================
+
+This table lists all supported client request headers.  Many of the
+headers can are updated in the client by response headers, and supplied
```

Comment:
can be updated in the client as response headers, and ..

Reply:
Changed.